src/transformers/modeling_blenderbot.py

```diff
@@ -0,0 +1,582 @@
+            .expand(batch_size, n_heads, query_len, key_len).view(batch_size * n_heads, query_len, key_len)
+        assert attn_mask.shape == dot_prod.shape
+        dot_prod.masked_fill(attn_mask, neginf(dot_prod.dtype))
+        attn_weight = F.softmax(dot_prod, dim=-1, dtype=torch.float).type_as(query)
+        attn_weight = self.att_dropout(attn_weight)
+        attention = attn_weight.bmm(v)
+        attention = (attention.type_as(query)).view(batch_size, n_heads, query_len, dim_per_head).transpose(1, 2) \
+            .contiguous().view(batch_size, query_len, dim)
+        output = self.out_lin(attention)
+        new_incremental_state = {
+            'prev_key': k.view(batch_size, n_heads, -1, dim_per_head),
+            'prev_value': v.view(batch_size, n_heads, -1, dim_per_head),
+            'prev_mask': mask
+        }
+        return  output, new_incremental_state
+
+
+class BlenderbotFFN(nn.Module):
+    """
+    implements the FFN
+    """
+
+    def __init__(self, dim, hidden_dim, relu_dropout=0.0, activation='relu'):
+        super(BlenderbotFFN, self).__init__()
+        self.relu_dropout = nn.Dropout(p=relu_dropout)
+        if activation == 'relu':
+            self.non_linear = F.relu
+        elif activation == 'gelu':
+            self.non_linear = gelu
+        else:
+            raise ValueError('Not defined activation function')
+        self.lin1 = nn.Linear(dim, hidden_dim)
+        self.lin2 = nn.Linear(hidden_dim, dim)
+
+        # weight initialization will be done using the pretrained model
+
+        # # initialise bias to zeros
+        # if self.lin1.bias is not None:
+        #     nn.init.zeros_(self.lin1.bias)
+        # if self.lin2.bias is not None:
+        #     nn.init.zeros_(self.lin2.bias)
+
+    def forward(self, input):
+        input = self.lin1(input)
+        input = self.non_linear(input)
+        input = self.relu_dropout(input)
+        input = self.lin2(input)
+        return input
+
+
+class BlenderbotEncoderLayer(nn.Module):
+    """
+    implemments a single layer of the encoder
+    """
+
+    def __init__(self, n_heads, embedding_size, ffn_size, 
+                 relu_dropout=0.0, attention_dropout=0.0, activation='relu', dropout=0.0, layer_norm_eps=1e-5):
+        super().__init__()
+        self.n_heads = n_heads
+        self.ffn_size = ffn_size
+        self.embedding_size = embedding_size
+        self.activation = activation
+        self.dropout = nn.Dropout(p=dropout)
+        self.norm1 = LayerNorm(self.embedding_size, eps=layer_norm_eps)
+        self.norm2 = LayerNorm(self.embedding_size, eps=layer_norm_eps)
+        self.attention = BlenderbotMultiHeadAttention(self.n_heads, self.embedding_size, dropout=attention_dropout)
+        self.ffn = BlenderbotFFN(self.embedding_size, ffn_size, relu_dropout=relu_dropout, activation=self.activation)
+
+    def forward(self, input_tensor, mask):
+        tensor_copy = input_tensor
```

Comment:
Don't really understand this line `tensor_copy`

Reply:
`tensor_copy` is a residual which is added to the output of each layer as in residual networks. Maybe it's better to rename it. Or do you think it's not needed at all?

But I made a mistake in the next line. the input to the attention layer shouldn't be `tensor_copy` but `input_tensor.`